File: binary-search/python/binary_search.py


def recursive_binary_search(array, target_value, start=0, end=None):
    if not end:
        end = len(array) - 1
    mid = int((start + end) / 2)

    if start > end:
        return -1
    if array[mid] == target_value:
        return mid
    elif array[mid] > target_value:
        return recursive_binary_search(array, target_value, start=0, end=mid-1)
    else:
        return recursive_binary_search(array, target_value, start=mid+1, end=end)


def binary_search(array, target_value):
    start = 0
    end = len(array) - 1
    while start <= end:
        mid = int((start + end) / 2)
        if array[mid] == target_value:
            return mid
        elif array[mid] > target_value:
            end = mid - 1
        else:
            start = mid + 1
    return -1

File: binary-search/scala/src/main/scala/BinarySearch.scala

  object BinarySearch {

    def binarySearch(n: Int, values: Array[Int]) ={
      def helper(acc: Array[Int], size: Int, start: Int): Int ={
        if (size == 0) -1
        else if(n == acc(size/2)) start + size/2
        else if(n > acc(size/2)) helper(acc drop (size/2+size%2),size - (size/2+size%2), start + (size/2+size%2))
        else helper(acc take (size/2+size%2),size - (size/2+size%2),start)
      }
      helper(values, values.length,0)
    }
  }